jmxtrans-output/jmxtrans-output-kafka/src/main/java/com/googlecode/jmxtrans/model/output/kafka/KafkaWriter2.java

```diff
@@ -0,0 +1,123 @@
+package com.googlecode.jmxtrans.model.output.kafka;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.collect.ImmutableMap;
+import com.googlecode.jmxtrans.exceptions.LifecycleException;
+import com.googlecode.jmxtrans.model.OutputWriterAdapter;
+import com.googlecode.jmxtrans.model.Query;
+import com.googlecode.jmxtrans.model.Result;
+import com.googlecode.jmxtrans.model.Server;
+import lombok.Getter;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.StringSerializer;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Map;
+
+import static com.fasterxml.jackson.databind.annotation.JsonSerialize.Inclusion.NON_NULL;
+
+public class KafkaWriter2 extends OutputWriterAdapter {
+    @Nonnull
+    @Getter
+    private final ObjectMapper objectMapper;
+    @Nonnull
+    @Getter
+    private final Map<String, Object> producerConfig;
+    @Nonnull
+    @Getter
+    private final String topic;
+    private Producer<String, String> producer;
+
+    public KafkaWriter2(@Nonnull ObjectMapper objectMapper, @Nonnull Map<String, Object> producerConfig, @Nonnull String topic) {
+        this.objectMapper = objectMapper;
+        this.producerConfig = producerConfig;
+        this.topic = topic;
+    }
+
+    @Override
+    public void start() throws LifecycleException {
+        producer = new KafkaProducer<String, String>(producerConfig);
```

Comment:
The OutputWriters are not meant to be restarted, so moving the initialization of the Kafka producer out of the writer is no problem on that ground.

The intent of splitting writer / factory has multiple facets (I'm happy to do a Google hagouts about it, if you want to chat a bit more about the details). Here are some of them:

* **splitting parsing / construction and execution:** Output writers were becoming messy, with too many responsibilities. Json parsing in the same class as the writer itself was cluttering the code, and make it harder to thing just about the execution of the writer or just about the parsing.
* **simpler lifecycle:** I really don't like that `start()` method in the `OutputWriter`. An object should be ready to be used once constructed. But since we usually have non trivial construction work, this is not well suited to be done in a constructor. Moving it to a factory makes the writer itself much simpler (at some poitn, I want to remove that `start()`method).
* **composability:** A lot of the writers can share a lot of components. Creating writers by composing existing classes is much easier if it is not done in the constructor ([prefer composition over inheritance applies here](https://en.wikipedia.org/wiki/Composition_over_inheritance)). As an example, the [`GraphiteWriterFactory`](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/GraphiteWriterFactory.java#L104-L128) can create both TCP or UDP based writers. And the UDP / TCP part is shared by multiple writers.